apiserver/facades/client/bundle/bundle.go

```diff
@@ -151,10 +156,111 @@ func (b *BundleAPI) GetChanges(args params.BundleChangesParams) (params.BundleCh
 }
 
 // ExportBundle exports the current model configuration as bundle.
-func (b *BundleAPI) ExportBundle() (params.StringResult, error) {
-	return params.StringResult{}, nil
+func (b *BundleAPI) ExportBundle() (params.BytesResult, error) {
+	if err := b.checkCanRead(); err != nil {
+		return params.BytesResult{}, common.ServerError(common.ErrPerm)
+	}
+
+	var exportConfig state.ExportConfig
+
+	// Skip attributes which are not required.
+	exportConfig.SkipActions = true
+	exportConfig.SkipCloudImageMetadata = true
+	exportConfig.SkipCredentials = true
+	exportConfig.SkipIPAddresses = true
+	exportConfig.SkipSSHHostKeys = true
+	exportConfig.SkipStatusHistory = true
+	exportConfig.SkipLinkLayerDevices = true
+
+	model, err := b.backend.ExportPartial(exportConfig)
+	if err != nil {
+		return params.BytesResult{}, common.ServerError(err)
+	}
+
+	err = model.Validate()
+	if err != nil {
+		return params.BytesResult{}, common.ServerError(err)
+	}
+
+	// Fill it in charm.BundleData datastructure.
+	bundleData, err := b.FillBundleData(model)
+	if err != nil {
+		return params.BytesResult{}, common.ServerError(err)
+	}
+
+	bytes, err := yaml.Marshal(bundleData)
+	if err != nil {
+		return params.BytesResult{}, common.ServerError(err)
+	}
+
+	return params.BytesResult{
+		Result: bytes,
+	}, nil
 }
 
 // Mask the new method from V1 API.
 // ExportBundle is not in V1 API.
 func (u *APIv1) ExportBundle() (_, _ struct{}) { return }
+
+// Fills the bundledata datastructure required for the exportBundle.
+func (b *BundleAPI) FillBundleData(model description.Model) (*charm.BundleData, error) {
```

Comment:
Any reason why this is an exported method?

It would be helpful if we had an interface declaration for v1 and v2 that would clearly highlight the differences between versions. This would also allow us to track differences in the future when we'd be at 3, 4, 5, etc

Reply:
I agree with @anastasiamac here, the method shouldn't be exported.